### ISSUE TITLE
Container CPU usage in MHZ

### ISSUE THREAD

**Issue by chidambaranathan-r** [google/cadvisor#2354]
![graph_5m](https://user-images.githubusercontent.com/1535689/70493126-a5a84200-1b2c-11ea-995f-e047c5412652.JPG)
Did a small study to report the CPU usage in terms of MHZ.

To get the POD’s cpu usage (sum of the CPU usage of each container belonging to the pod), the below query is used that will give the number of cores that are being used by each pod.

sum by (pod_name) (rate (container_cpu_usage_seconds_total{namespace="default", name=~".*cpu-stress.*"}[5m]))

Attached the sample graph for reference. CPU simulation is done by deploying a sample deployment that uses stress tool to stress the CPU.
		
As in the above query, the final output per pod will be in terms of cores. To convert the cores into MHZ, there is no out of the box solution from Prometheus (through cAdvisor) even with queries as I understood.

As per the study, there would be a manual intervention required to convert the cores into MHZ because the meaning of CPU differs in various platform. Challenge here is to understand whether one CPU in Kubernetes is one physical core or one Hyper-thread and what is the corresponding clock speed? It varies based on the infrastructure that is being used to run Kubernetes.

One CPU, in Kubernetes, is equivalent to:
•	1 AWS vCPU
Each vCPU is a thread of a CPU core, except for T2 instances. For T2 instances, 1 vCPU = 1 physical core. For all others, 1 vCPU = 1 logical core.
Reference: https://docs.aws.amazon.com/AWSEC2/latest/UserGuide/instance-optimize-cpu.html ; https://aws.amazon.com/ec2/instance-types/
•	1 GCP Core
A vCPU is implemented as a single hardware Hyper-thread on one of the available CPU platforms.
Reference: https://cloud.google.com/compute/docs/machine-types ; https://cloud.google.com/compute/docs/cpu-platforms
•	1 Azure vCore
1 vCore = 1 hyper-thread, except for Gen4 hardware generation.
Reference: https://docs.microsoft.com/en-us/azure/sql-database/sql-database-service-tiers-vcore
•	1 Hyperthread on a bare-metal Intel processor with Hyperthreading

Let us take the bare-metal case as an example:
Hyper-thread: enabled
CPU: 2 sockets with 24 cores each (48 physical cores), yielding 96 Hyper-threads.
Clock-speed of each core: 2.7GHZ or (2700MHZ)
So, the clock-speed of single hyper-thread would be 2700MHZ / 2 threads = 1350MHZ
If in case, POD has used 100m core then the it can be converted as below to show in MHZ,
(100 * 0.001) * 1350MHZ = 135 MHZ

Any other solution in my scenario please? 

Also, I feel that this could be facilitated by cadvisor itself to have a metrics for CPU in MHZ. Do you agree?

**Comment by dashpole** [comment]
Is the suggestion that cAdvisor should be aware of the clock speeds of vCPUs provided in different clouds?  It seems like that would be very difficult to keep up-to-date...

**Comment by chidambaranathan-r** [comment]
Yeah. At-least if cadvisor could able to give the clock speed as a separate metric, that would be good where we can use it as part of query to derive the CPU usage in MHZ.

Just a thought: Is it not possible to get from the OS (ex: lscpu etc.) or any other form and report the same?

I referred this blog: https://docs.vmware.com/en/VMware-vSphere/6.5/com.vmware.vsphere.monitoring.doc/GUID-FBA3A58A-EEC9-4B4D-8515-ECCF51188913.html

But not sure how are they doing it.

Please share your thoughts.

**Comment by dashpole** [comment]
I don't think that really belongs in cAdvisor.  It might be more appropriate in something like the [node_exporter](https://github.com/prometheus/node_exporter), which could then be multiplied by the cAdvisor CPU usage metric.

**Comment by chidambaranathan-r** [comment]
Thanks for sharing your views. Will look into that.

One last query:
I just want to understand the theory behind the below:

When 'container_cpu_usage_seconds_total' is viewed without any function applied over it, it shows the output in 'seconds'.

Whereas, when rate function is applied over it, it shows the output in 'number of cores'.

Maybe its simple. Please clarify.

**Comment by dashpole** [comment]
The cumulative metric is an amount of time spent running on cpu cores, measured in seconds.  We sometimes refer to this as, for example, 10000 core-seconds, since it is that amount time spent on a cpu core.

Taking a rate just divides by time, giving core-seconds / second, or the average rate of usage of CPU cores.

**Comment by chidambaranathan-r** [comment]
Thanks for your clarification. Really appreciate your speedy response in both of my cases.

**Comment by dims** [comment]
Closing this issue as it has been open since before January 1, 2024. If this is still relevant, please feel free to reopen or create a new issue.
